rescan_test.go

```diff
@@ -504,3 +504,188 @@ func (ctx *rescanTestContext) waitUntilSynced() {
+	// successfully.
+	ctx.chain.setFailGetFilter(false)
+
+	// The blocks should now be notified in their expected order.
+	ctx.assertFilterQueried(block1.Hash)
+	ctx.recvBlockConnected(block1)
+	ctx.assertFilterQueried(block2.Hash)
+	ctx.recvBlockConnected(block2)
+}
+
+// TestRescanReorgAllButFirstRetryBlocks ensures the rescan stops retrying
+// blocks for which it has received disconnected notifications for.
+func TestRescanReorgAllButFirstRetryBlocks(t *testing.T) {
+	t.Parallel()
+
+	// Dummy options to allow the Rescan to fetch block filters.
+	ctx := newRescanTestContext(t, 10, []RescanOption{
+		StartTime(time.Time{}),
+		WatchInputs(InputWithScript{}),
+	}...)
+	ctx.start()
+	defer ctx.stop()
+
+	ctx.waitUntilSynced()
+
+	// Modify the mocked chain such that it fails to retrieve block filters.
+	ctx.chain.setFailGetFilter(true)
+
+	// Adding multiple new blocks to the chain should result in the first
+	// retry block being queried until successful.
+	block := ctx.chain.addNewBlock(true)
+	ctx.assertFilterQueried(block.Hash)
+	ctx.chain.addNewBlock(true)
+	ctx.assertFilterQueried(block.Hash)
+	ctx.chain.addNewBlock(true)
+	ctx.assertFilterQueried(block.Hash)
+
+	// We'll then reorg back to the first retry block. This should cause us
+	// to stop retrying the following reorged blocks.
+	ctx.chain.rollbackToHeight(block.Height, true)
+
+	// Revert the mocked chain so that block filters can be retrieved
+	// successfully.
+	ctx.chain.setFailGetFilter(false)
+
+	// Since the first block wasn't reorged, we should expect a notification
+	// for it.
+	ctx.assertFilterQueried(block.Hash)
+	ctx.recvBlockConnected(block)
+}
+
+// TestRescanReorgAllRetryBlocks ensures the rescan stops retrying blocks for
+// which it has received disconnected notifications for and queries for blocks
+// of the new chain instead.
+func TestRescanReorgAllRetryBlocks(t *testing.T) {
+	t.Parallel()
+
+	// Dummy options to allow the Rescan to fetch block filters.
+	ctx := newRescanTestContext(t, 10, []RescanOption{
+		StartTime(time.Time{}),
+		WatchInputs(InputWithScript{}),
+	}...)
+	ctx.start()
+	defer ctx.stop()
+
+	ctx.waitUntilSynced()
+
+	// Note the block which we'll revert back to.
+	bestBlock, err := ctx.chain.BestBlock()
+	if err != nil {
+		ctx.t.Fatalf("unable to retrieve best block: %v", err)
+	}
+
+	// Modify the mocked chain such that it fails to retrieve block filters.
+	ctx.chain.setFailGetFilter(true)
+
+	// Adding multiple new blocks to the chain should result in the first
+	// retry block being queried until successful.
+	block1 := ctx.chain.addNewBlock(true)
+	ctx.assertFilterQueried(block1.Hash)
+	block2 := ctx.chain.addNewBlock(true)
+	ctx.assertFilterQueried(block1.Hash)
+
+	// We'll then reorg back to a height where the new blocks we've added
+	// and retried are stale, causing us to no longer retry them. This won't
+	// deliver any disconnected notifications since the blocks were never
```

Comment:
I think the comment should be slightly changed to point out that they were never handled from the POV of the rescan. Disconnected notifications are delivered from the chain source I reckon?

Reply:
Done.